src/hotspot/share/memory/iterator.hpp

```diff
@@ -41,7 +41,7 @@ class Thread;
 
 // The following classes are C++ `closures` for iterating over objects, roots and spaces
 
-class Closure : public StackObj { };
+class Closure : public CHeapObj<mtGC> { };
```

Comment:
Why is this required (in this PR)?

Reply:
The changes need to allocate `G1CollectedHeap::_is_alive_closure_cm` when creating that argument for creating the concurrent reference processor. Previously the closure used `G1CollectedHeap` which is available when inlining it into the `G1CollectedHeap` instance, but now the closure needs `G1ConcurrentMark` which is not available at that time (i.e. null).
The alternative would have been introducing something like an `initialize` method to that closure which I did not like.

Reply:
OK, could `G1CMIsAliveClosure::do_object_b` retrieve `_cm` via `_g1h`? (It's obvious why `Closure` is part of mtGC here.)

Reply:
At the cost of an additional indirection per reference. I would prefer to just have an `initialize` method then.
You are right that just using `mtGC` is not appropriate here as it is used in non-GC code too. Is an `initialize` method okay with you then?